### ISSUE TITLE
False warning for S2583 in try/catch

### ISSUE THREAD

**Issue by jcurl** [SonarSource/sonar-csharp#1692]
### Description

S2583 false warning is generated for try/catch blocks

### Repro steps

The following code snippet produces a false warning (the line in the Exception block).

```csharp
public static MyVar Check() {
  MyVar var = null;
  try {
    var = Factory.Create();
    var.Check();
    return var;
  } catch (Exception) {
    if (var != null) var.Dispose();
    throw
  }
}
```

### Expected behavior

A S2583 should not arise, as the variable `var` can very well be `null`, if the method `Factory.Create()` raises an exception. But it won't be `null` if an exception is raised in `var.Check()`. Thus, in case of an exception, the code wants to clean up and reraise the exception.

### Actual behavior

S2583 is raised for the portion `var != null` in the exception block.

### Known workarounds

No known workarounds.

### Related information

* SonarC# Version (3.6.0.2584 SonarLint for Visual Studio 2015)
* Visual Studio Version 2015

**Comment by valhristov** [comment]
Hi @jcurl, this problem occurs because the analyzer does not follow all possible execution paths in the try-catch block and it cannot detect that the variable has been actually initialized somewhere within the try block.

To avoid the false positive we [recently changed](https://github.com/SonarSource/sonar-csharp/pull/1490) the rule to ignore always true/false conditions that are inside catch or finally block. The change went live a few months ago and if you upgrade your SonarLint to the latest version, the problem should disappear. Note that since version 4.0, SonarLint is compatible with Visual Studio 2015 Update 3 and newer.
